src/hal/src/external_memory/mod.rs

```diff
@@ -0,0 +1,498 @@
+    pub fn handle(&self) -> Option<&Handle> {
+        match self {
+            Self::OpaqueWin32(handle) => Some(handle),
+            Self::OpaqueWin32Kmt(handle) => Some(handle),
+            Self::D3D11Texture(handle) => Some(handle),
+            Self::D3D11TextureKmt(handle) => Some(handle),
+            Self::D3D12Heap(handle) => Some(handle),
+            Self::D3D12Resource(handle) => Some(handle),
+            _ => None,
+        }
+    }
+
+    /// Get the host pointer of this external memory
+    pub fn ptr(&self) -> Option<&Ptr> {
+        match self {
+            Self::HostAllocation(ptr) => Some(ptr),
+            Self::HostMappedForeignMemory(ptr) => Some(ptr),
+            // Without this on non unix or windows platform, this will trigger error for unreachable pattern
+            #[allow(unreachable_patterns)]
+            _ => None,
+        }
+    }
+}
+
+impl From<ExternalMemory> for (ExternalMemoryType, PlatformMemory) {
+    fn from(external_memory: ExternalMemory) -> (ExternalMemoryType, PlatformMemory) {
+        match external_memory {
+            #[cfg(unix)]
+            ExternalMemory::OpaqueFd(fd) => (ExternalMemoryType::OpaqueFd, fd.into()),
+            #[cfg(windows)]
+            ExternalMemory::OpaqueWin32(handle) => (ExternalMemoryType::OpaqueWin32, handle.into()),
+            #[cfg(windows)]
+            ExternalMemory::OpaqueWin32Kmt(handle) => {
+                (ExternalMemoryType::OpaqueWin32Kmt, handle.into())
+            }
+            #[cfg(windows)]
+            ExternalMemory::D3D11Texture(handle) => {
+                (ExternalMemoryType::D3D11Texture, handle.into())
+            }
+            #[cfg(windows)]
+            ExternalMemory::D3D11TextureKmt(handle) => {
+                (ExternalMemoryType::D3D11TextureKmt, handle.into())
+            }
+            #[cfg(windows)]
+            ExternalMemory::D3D12Heap(handle) => (ExternalMemoryType::D3D12Heap, handle.into()),
+            #[cfg(windows)]
+            ExternalMemory::D3D12Resource(handle) => {
+                (ExternalMemoryType::D3D12Resource, handle.into())
+            }
+            #[cfg(any(target_os = "linux", target_os = "android"))]
+            ExternalMemory::DmaBuf(fd) => (ExternalMemoryType::DmaBuf, fd.into()),
+            #[cfg(target_os = "android")]
+            ExternalMemory::AndroidHardwareBuffer(fd) => {
+                (ExternalMemoryType::AndroidHardwareBuffer, fd.into())
+            }
+            ExternalMemory::HostAllocation(ptr) => (ExternalMemoryType::HostAllocation, ptr.into()),
+            ExternalMemory::HostMappedForeignMemory(ptr) => {
+                (ExternalMemoryType::HostMappedForeignMemory, ptr.into())
+            }
+        }
+    }
+}
+
+impl std::convert::TryFrom<(ExternalMemoryType, PlatformMemory)> for ExternalMemory {
+    type Error = &'static str;
+    fn try_from(tuple: (ExternalMemoryType, PlatformMemory)) -> Result<Self, Self::Error> {
```

Comment:
Do we have to use `TryFrom` here? It seems to me that we are trying to fit into this trait for no apparent benefit.
Just having a method `fn from_platform(ty: ExternalMemoryType, pm: PlatformMemory) -> Result<Self, &'static str>` would be cleaner.

Reply:
Effectively your approach is much better and cleaner. Done!